detectors/unprotected-mapping-operation/src/lib.rs

```diff
@@ -0,0 +1,202 @@
+}
+
+struct AuthStatus {
+    authed: bool,
+}
+
+struct UnauthorizedAddress {
+    span: Span,
+    name: String,
+}
+
+struct UnprotectedMappingOperationFinder<'tcx, 'tcx_ref> {
+    cx: &'tcx_ref LateContext<'tcx>,
+    linked_addresses: Vec<(AuthStatus, Vec<HirId>)>,
+    unauthorized_span: Vec<UnauthorizedAddress>,
+}
+
+const SOROBAN_MAP: &str = "soroban_sdk::Map";
+const SOROBAN_ADDRESS: &str = "soroban_sdk::Address";
+
+impl<'tcx> LateLintPass<'tcx> for UnprotectedMappingOperation {
+    fn check_fn(
+        &mut self,
+        cx: &LateContext<'tcx>,
+        _: rustc_hir::intravisit::FnKind<'tcx>,
+        _: &'tcx rustc_hir::FnDecl<'tcx>,
+        body: &'tcx rustc_hir::Body<'tcx>,
+        _: Span,
+        _: rustc_span::def_id::LocalDefId,
+    ) {
+        let mut visitor = UnprotectedMappingOperationFinder {
+            cx,
+            linked_addresses: Vec::new(),
+            unauthorized_span: Vec::new(),
+        };
+
+        visitor.parse_body_params(body.params);
+
+        walk_body(&mut visitor, body);
+
+        visitor
+            .unauthorized_span
+            .iter()
+            .for_each(|unauthorized_address| {
+                Detector::UnprotectedMappingOperation.span_lint_and_help(
+                    cx,
+                    UNPROTECTED_MAPPING_OPERATION,
+                    unauthorized_address.span,
+                    &format!(
+                        "Address not authorized, please use `{}.require_auth();` prior to the mapping operation",
+                        unauthorized_address.name
+                    ),
+                );
+            });
+    }
+}
+
+impl<'tcx> Visitor<'tcx> for UnprotectedMappingOperationFinder<'tcx, '_> {
+    fn visit_expr(&mut self, expr: &'tcx Expr<'tcx>) {
+        if let ExprKind::Block(block, _) = &expr.kind {
+            block.stmts.iter().for_each(|stmt| {
+                if_chain! {
+                    if let StmtKind::Local(local) = &stmt.kind;
+                    if self.get_node_type(local.hir_id) == SOROBAN_ADDRESS;
+                    if let PatKind::Binding(_, target_hir_id, _, _) = &local.pat.kind;
+                    if let Some(init) = &local.init;
+                    let source_hir_id = self.get_expr_hir_id(init);
+                    if let Some(source_hir_id) = source_hir_id;
+                    then {
+                        // Insert the new address into the linked_addresses
+                        self.insert_new_address(source_hir_id, *target_hir_id);
+                    }
+                }
+            })
+        }
+
+        if let ExprKind::MethodCall(method_path, method_expr, method_args, _) = &expr.kind {
+            // Get the method expression type and check if it's a map with address
+            let method_expr_type = self.get_node_type(method_expr.hir_id);
+
+            if method_expr_type.starts_with(SOROBAN_MAP)
+                && method_expr_type.contains(SOROBAN_ADDRESS)
+            {
```

Comment:
Is it relevant that now if the Address is used as Value in the map, the detector triggers?